controllers/configuration/backend/s3.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backend
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/oam-dev/terraform-controller/api/v1beta2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	s3AccessKey    = "AWS_ACCESS_KEY_ID"
+	s3SecretKey    = "AWS_SECRET_ACCESS_KEY"
+	s3SessionToken = "AWS_SESSION_TOKEN"
+)
+
+// S3Backend is used to interact with the Terraform s3 backend
+type S3Backend struct {
+	client       s3iface.S3API
+	AccessKey    string
+	SecretKey    string
+	SessionToken string
+	Region       string
+	Key          string
+	Bucket       string
+}
+
+func newS3Backend(ctx context.Context, options *buildBackendOptions) (Backend, error) {
+	conf, ok := options.backendConf.(*v1beta2.S3BackendConf)
+	if !ok || conf == nil {
+		return nil, fmt.Errorf("invalid backendConf, want *v1beta2.S3BackendConf, but got %#v", options.backendConf)
+	}
+	s3Backend := &S3Backend{
+		Region: conf.Region,
+		Key:    conf.Key,
+		Bucket: conf.Bucket,
+	}
+	if err := s3Backend.fillOptions(ctx, options.k8sClient, options.configurationNS, options.extraOptionSource); err != nil {
+		return nil, err
+	}
+	if err := s3Backend.buildClient(); err != nil {
+		return nil, err
+	}
+	return s3Backend, nil
+}
+
+func (s *S3Backend) fillOptions(ctx context.Context, k8sClient client.Client, namespace string, optionSource *OptionSource) error {
+	accessKey, ok, err := optionSource.getOption(ctx, k8sClient, namespace, s3AccessKey)
```

Comment:
Can we use the same method to get the credentials https://github.com/loheagn/terraform-controller/blob/99eb18e5a8a5c1e752730b502187fcdf89a10fbc/controllers/provider/credentials.go#L60?

Reply:
Yes. I can update the code to reuse the `getAWSCredentials` function.